Add upper and lower limits of temperature in phase transitions

lhy11009 opened the pull request:
Pull Request Checklist. Please read and check each box with an X. Delete any part not applicable. Ask on the [forum](https://community.geodynamics.org/c/aspect) if you need help with any step.

I added temperature limits on phase transition so as to only activate them within a certain range of temperature. The default values for them are -1.0 (negative values mean limits are not applied). This feature is useful to apply a composition-dependent phase transitions as some PTs have distinct boundaries in hotter & colder regions (e.g. the basalt -> eclogite; gt -> pv). Previously, I made hard modifications in the source code to facilitate the phase transitions for pyrolite and basalt, but I recently came up with this solution within our current structure of phase transitions.


I have added 3 tests to demonstrate the usage in the case of pyrolite, harzburgite, and basalt composition, respectively.
If @gassmoeller or @jdannberg could take a look, that will be appreciated. The implementation in the "material_model/utilities.cc" file is a little messy for now. I would like to hear your opinion on that. @mibillen would also be involved in the conversation. Moreover, I'll also notify @MFraters and @rfildes that this is being handled, so that they could use my implementation even before it's finalized.

### Before your first pull request:

* [x] I have read the guidelines in our [CONTRIBUTING.md](../blob/main/CONTRIBUTING.md) document.

### For all pull requests:

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

lhy11009 commented:
Here is an illustration of the density variations: A) Pyrolite, B) Harzburgite, C) Basalt. These come from the visualization of the three tests I enlisted here.

![densities-01](https://user-images.githubusercontent.com/28665365/168453643-7f09fd1a-13d6-4b3b-b268-b6fc8f24df9d.png)

jdannberg reviewed (commented):
I think the general approach makes sense. I really like that this allows you to prescribe different phase transitions for different temperatures, which I always thought was a big limitation of the phase function approach. I have two concerns that we should discuss

1. In the current implementation of `phase_average_value` in utilities.cc, the material properties at a given location are computed as `averaged_parameter += phase_function_values[base-composition+i] * (parameter_values[base+i+1] - parameter_values[base+i])`. So imagine the following scenario: you have two phase transitions, and they are supposed to occur at similar pressures, but one only occurs at a low temperature, and the other one at a high temperature. In the input file, you list the one at low temperature first, then the one at high temperature. The density is supposed to be 3000 above, and 3500 below the transitions. So it would be intuitive to set the densities to `set Densities = background:3000|3500|3500`. For the transition at low temperature, you get the right density jump: you multiply the phase function by (3500-3000), and add this to the base value of 3000. But for the high-temperature transition, this would not actually work, since the phase function of the low-temperature transition would be zero in the high-temperature region, so you would compute the density as base value + phase function (high-temperature transition) * (3500-3500) = 3000. You could change the values to make it work correctly, but it would be very confusing, because the density values in the input file would not correspond to the actual densities in the model.  
2. The current approach also allows you to create very inconsistent phase diagrams. Imagine you only have a phase transition for a lower temperature range, but no phase transition for a higher temperature range. If material circulates, moving down at low temperatures and moving up at high temperatures, it would only encounter the phase transition on the way down. So, in the case of an exothermic phase transition, you could introduce unlimited energy into the model. I do not see a good way to fix this, but I think it would be worth documenting that one should be careful to be consistent with the definition of the different phase transitions to avoid this scenario.

lhy11009 commented on source/material_model/utilities.cc:
```diff
@@ -1199,6 +1221,20 @@ namespace aspect
+                           "upper limit of temperature for the specific phase transition. "
+                           "List must have the same number of entries as Phase transition depths. "
+                           "Units: \\si{\\kelvin}.");
+        prm.declare_entry ("Phase transition temperature lower limits", "-1.0",
```
Yes, I think these also make a lot more sense to me.

lhy11009 commented on tests/visco_plastic_phases_basalt.prm:
```diff
@@ -0,0 +1,102 @@
+# Copy of visco_plastic.prm, but include a phase transition for the EoS
```
I have added this information to every file.

lhy11009 commented on tests/visco_plastic_phases_basalt/screen-output:
```diff
@@ -0,0 +1,29 @@
+-----------------------------------------------------------------------------
```
I have appended this output to each of the test.

lhy11009 commented on tests/visco_plastic_phases_harzburgite.prm:
```diff
@@ -0,0 +1,102 @@
+# Copy of visco_plastic.prm, but include a phase transition for the EoS
```
Updated.

lhy11009 commented on doc/modules/changes/20220511_haoyuan:
```diff
@@ -0,0 +1,4 @@
+New: The lower and upper limits of temperature of phase transitions.
```
same here.

lhy11009 replied:
I have manually changed the file (The commit suggestion option is grayed out on this page).

lhy11009 commented:
Hi @jdannberg, thanks for your valueble reviewing. I have addressed all of them in the source code. Moreover, I found your questions really hitting the heart of the issue. I'll try to answer them here.

To start with, it's actually how we think of the phase transitions. So the point I want to make is, they should be thought as transitions but not necessarily as discrete phases. For example, in the visualization I included for the basalt composition, there are actually 3 distinct transitions for the basalt -> eclogite (light yellow to purple). On one hand, this thought enlightens me into implementing the temperature limits into the source coed. On the other hand, this goes into how as set up the parameters: I set them as jumps in density and viscosity.

Then I'll follow this consideration to answer your questions. In the first case you made, the "higher temperature" transition would require another jump of "500" in the density. So these are two distinct PZs between just 2 phases. In the second case you made, we could have another phase transition right at the boundary of the "lower" and "higher" temperature ranges . It's acctually what I did for the small vertical segment in the basalt composition (yellow to purple, around 500 C ) - a narrow range in temperature set by the "limits" of temperature with a very large claypeyron slope.

Thirdly, I'll follow your consideration about the confusiion in the parameters. For the basalt, I assigned this:
`set Heat capacities = background:3000.0|3540.0|4080.0|4620.0|4693.0|4951.7`
The differences in the 1st, 2nd and 3rd phase transitions are all 540, in this way, they don't show the real density in the phases. For sure, this is confusing, but it will work for a variaties of things. What makes more sense to me would be to modify the interfaces in this way: entering a base value and variations across PZs rather than the real value for each phases. That way, the confusing part is lifted as well, where I would be able to enter three "540" into this variable, instead of "3000.0", "3540.0" and "4080.0". However, that may deprecate the previous implementation.

Last, there might be issue with this approach, if all the segments of PZs across two phases don't cover a whole range of temperature in the model, so it would always require first plot out the phase diagram following the "phase diagram" cookbook and check both densities and viscosities.

lhy11009 commented:
@jdannberg There seems to be a weird bug with the line we added for the maximum value in "utilities.cc". I don't have this issue building it on my local machine.

bangerth commented on doc/modules/changes/20220511_haoyuan:
```diff
@@ -0,0 +1,4 @@
+New: The lower and upper limits of temperature of phase transitions.
+Phase transitions could be actived within a temperature range, and deactived outside.
+<br>
+(Haoyuan Li, 2022/05/11)
```
github complains that this line is missing a newline. Can you add one?

jdannberg commented:
Do you include `#include <boost/lexical_cast.hpp>` in the file where you declare the parameter?

I don't know if that is actually what causes the problem, but I made the suggestion based on the "Global composition maximum" parameter in parameters.cc, and that includes this. (And if you use a unity build, it might happen that some of the problems due to not including specific files do not show up, but the tester will then find them.)

jdannberg commented:
Thank you for making these changes. I think there are still some fundamental points we need to discuss and fix before this can be merged. 

(1) Using densities (rather than density jumps) makes the input files very confusing. This is not something we can merge because it is not intuitive to use. As you point out, using density jumps would break all existing input files, and many people use the multicomponent equation of state. 

(2) The densities should be stored in the densities rather than in the heat capacities. If that doesn’t work for some reason, we should fix it, but this is very confusing for a test case.

(3) Can we think about any safeguards, or at least sufficient documentation, to make sure that users do not create phase diagrams that make no sense? Maybe there is no perfect solution, but we should at least do the best we can.

I don’t have a good idea for how to fix these three points, so it might be worth to have an in-person/zoom discussion with more people to brainstorm ideas (like, make using density jumps an option? Making a new equation of state? Figuring out if there is a less confusing way of using densities? etc).

jdannberg commented:
I am not sure there will be a meeting on the 24th (many of us will be on our way back from the hackathon). But the next meeting after that would work.

lhy11009 commented:
@jdannberg Okay, Let's plan on discussing it there. If it comes down to implementing a new EOS, I think I could alos handle it with your help. So, I guess we'll have to close this PR for now.

lhy11009 commented:
Some updates on the example of the basalt phase transitions, on top of this other PR: https://github.com/geodynamics/aspect/pull/4868
Left: density profile. Right: viscosity profile
![basal_phases_combined](https://user-images.githubusercontent.com/28665365/222864748-c7e05658-062a-469d-8a68-c8ebdd676bab.jpg)

lhy11009 commented:
During the recent rebase to the main branch, I made some changes to the comments and tests:

1. I added the comment to the options of "Phase transition temperature lower limits" and Phase transition temperature upper limits" to tell users to be careful about these options.
2.  I have fixed the test to produce the designated results for this PR. But in the context of this PR, it's hard to prescribe the values that actually make sense regarding the jump in density (as @jdannberg has mentioned). And I also use the specific heat to demonstrate the change in the interface.  So If that makes sense to the reviewers, we can keep these 3 tests just as testing the interface rather than making models with physical meanings. And in a future Cookbook, I'll demonstrate the usage of these interfaces in a physical model.

lhy11009 commented:
@gassmoeller @jdannberg @mibillen Can we handle the review of this PR first? In this way, I'll be able to work on the cookbook in the other PR as well. Thanks.

gassmoeller reviewed (commented):
Thanks for the updates. This is very close to ready. I have some suggestions about wording and a request for a change in the test resolution, but then this PR should be good to go. While the current way of inputting values is not ideal that is beyond the scope of this PR and will be addressed in #4868.

lhy11009 commented on source/material_model/utilities.cc:
```diff
@@ -1251,6 +1267,29 @@ namespace aspect
+                           boost::lexical_cast<std::string>(std::numeric_limits<double>::max()),
+                           Patterns::Anything(),
+                           "A list of upper limits of temperature, determining a range where the "
+                           "the phase transition is activated. A negative value means there is no "
+                           "upper limit of temperature for the specific phase transition. "
```
I have applied the change of using the lower limit of float number here.

gassmoeller replied:
You must have lost the commit that contained these changes, this line is unchanged to when I wrote my earlier comment. This appears at several other places as well, I will mark them. As a general rule of thumb I always read through my own pull request on github again after uploading it, to make sure I have uploaded the right version. You can always see all of your changes under the "Files changed" tab (e.g. https://github.com/geodynamics/aspect/pull/4591/files). Reading through them yourself and fixing them as soon as you notice something often avoids another round of review.

lhy11009 commented on tests/visco_plastic_phases_temperature_limit_basalt.prm:
```diff
@@ -0,0 +1,110 @@
+subsection Geometry model
+  set Model name = box
+  subsection Box
+    set X repetitions = 50
+    set Y repetitions = 50
```
I have reduced the meshes for all the 3 tests. The first lines are changes as well.

gassmoeller replied:
same here, we are again at the old state of 50x50. Same for the other tests

lhy11009 reviewed (commented):
I have applied all the suggested changes. Every test seems to be working as well. Can you take a look ? @gassmoeller . Also, can you kindly remind me what's the current method to compile the manual? I used to make the manual.pdf under /doc folder but that doesn't seem to work right now.

gassmoeller commented on doc/Makefile:
```diff
@@ -1,132 +1,223 @@
-############################################################
```
You probably didnt mean to delete this file.

gassmoeller commented on doc/modules/changes/20220511_haoyuan:
```diff
@@ -0,0 +1,4 @@
+New: The lower and upper limits of temperature of phase transitions.
+Phase transitions could be actived within a temperature range, and deactived outside.
+<br>
```
There is still an unresolved comment by jdannberg about this changelog entry. For reference, this is the comment: https://github.com/geodynamics/aspect/pull/4591/files#r874136668

lhy11009 replied:
Hi Rene, I have merged your PR and fixed the tests with the 3 new prm files.

gassmoeller reviewed (commented):
Thanks, except for my two comments this is now ready.

Also take a look at https://github.com/lhy11009/aspect/pull/2. We can handle this as a separate PR after this one is merged.

lhy11009 reviewed (commented):
I have addressed the comments and merged your PR as well as fixed the tests. I see that I didn't make any changes to the doc/Makefile, and I am not sure what it is. I'll check again if there is a problem.

gassmoeller commented on source/material_model/utilities.cc:
```diff
@@ -1251,6 +1267,29 @@ namespace aspect
+                           "the phase transition is activated. A negative value means there is no "
+                           "upper limit of temperature for the specific phase transition. "
+                           "List must have the same number of entries as Phase transition depths. "
+                           "When the options of temperature limits are adapted, the user has to be "
+                           "careful about the consistency between adjacent phases to make perfect "
+                           "sense in their physics. We recommend producing a phase diagram with "
+                           "simple model setups to check the implementation as a starting point."
```
This change has not been applied.

gassmoeller commented on source/material_model/utilities.cc:
```diff
@@ -1251,6 +1267,29 @@ namespace aspect
+                           "sense in their physics. We recommend producing a phase diagram with "
+                           "simple model setups to check the implementation as a starting point."
+                           "Units: \\si{\\kelvin}.");
+        prm.declare_entry ("Phase transition temperature lower limits", "0.0",
```
This change has not been applied.

gassmoeller commented on source/material_model/utilities.cc:
```diff
@@ -1251,6 +1267,29 @@ namespace aspect
+                           "the phase transition is activated. A negative value means there is no "
+                           "lower limit of temperature for the specific phase transition. "
+                           "List must have the same number of entries as Phase transition depths. "
+                           "When the options of temperature limits are adapted, the user has to be "
+                           "careful about the consistency between adjacent phases to make perfect "
+                           "sense in their physics. We recommend producing a phase diagram with "
+                           "simple model setups to check the implementation as a starting point."
```
not committed

gassmoeller reviewed (commented):
You must have deleted the commit that addressed my earlier changes during a rebase or squash, all of my earlier comments are not addressed. Please take another look.

lhy11009 commented:
Sorry about messing it up. This time I am more careful and came across a potential issue in the new implementation. I am trying to depict this from the following diagrams of the pyrolite composition. The subfigure on the left is the test with the old implementation and the subfigure on the right is the test with the new implementation. Can you tell the differences? While the change is the "rd-> pv +mw" transition. In this case, the subfigure on the left maintains this transition at colder temperatures, while the subfigure on the right doesn't. I have an idea that this is from the sequence of the phase transitions on 660. 
![combine_images (3)](https://user-images.githubusercontent.com/28665365/228420293-31d8c836-98a7-45c2-a6b6-05f3b6372f53.jpg)
Let's gather at tomorrow's meeting and work on it.

lhy11009 commented:
Hi Rene.
I have tried to fix both of these tests. In the first case, I apply a different phase transition for rd->pv+mw below 1662.0. This fixed the phase transition below 1662.0. The result is the subfigure on the left.
However, if I migrate the phase transition of rd-pv+mw to a cooler temperature, the fix is not exact (the right subfigure). Let me know what you think about it. Also if you have questions about the formulas I put in the other PR, feel free to tell me. We can schedule meeting individually if needed.

![pyrolite_fix_combine](https://user-images.githubusercontent.com/28665365/228717930-a2c54003-4d79-4b31-8df0-e6536b768244.jpg)

gassmoeller commented:
> However, if I migrate the phase transition of rd-pv+mw to a cooler temperature, the fix is not exact (the right subfigure). 

Can you describe a bit more what you mean by "not exact"? Is the goal that the right figure looks exactly like the left? And which part of the figure is not like what you expected? (i.e. does the purple phase start in the correct pressure-temperature conditions? does the black phase start correctly? is the black phase correct below 1662 K but wrong above?)


Separately: @jdannberg and I had a discussion about the different ways to write phase functions, and I think we came to the conclusion that at least in regions where several phase functions overlap, any definition of a simple phase function is not uniquely defined. E.g. consider the following case: we have phases P1,P2,P3 and phase functions `a` (P1->P2) and `b` (P2->P3). Assume a = 0.8 and b = 0.2. Then we could have two proportions of phases:

1. P1 = 0, P2 = 0.8, P3 = 0.2 (this means the 'b' transition uses up P1)
2. P1 = 0.2, P2 = 0.6, P3 = 0.2 (this means the 'b' transition only converts from P2)
3. P1 = 0.16, P2 = 0.64, P3 = 0.2 (this means the 'b' transition proportionally converts P1 and P2 into P3)

From our definition of phase functions it is impossible to say which scenario is intended. So I think the goal for our PR should be to find a formulation that is correct for all scenarios where only one phase function is not significantly different from 0 or 1.

lhy11009 commented:
Hi Rene

What I mean by "not exact". I am sorry that I don't know how to answer that straight enough. I  have to start from scratch.

lhy11009 commented:
Hi Rene

I just get back from a walk and I have a more general way to illustrate the issue I want to address with the previous tests (following figure).
So with two PTs crossing each other (PT2, PT3), we already know that these two need to be split with the intercepting point. 
Let's see do you agree that we need to further split every phase transition along the same value of temperature?
In this case, if I want to really stay on the safe side, I would want to split PT1 along the "split line" defined by the intercepting point of PT2 and PT3.
The reason is that I don't if PT1 has a definite relation (i.e. it doesn't intercept) with all the other phases.
And in the diagram I drew, it's very possible that PT1 would intercept with PT2 at a hotter temperature.

![image](https://user-images.githubusercontent.com/28665365/230798127-00976f7d-06f4-4470-b1fd-9ded99a5548e.png)

lhy11009 commented:
1. the current tests are working pretty well.
2. This thought helps me in prescribing the phase transitions. At first, they look complex and hard to figure out, but when I split them into subgroups (for one composition) by the temperature limit, for example:
`        set Phase transition temperature lower limits = background:-1e31|769.0|773.15|  -1e31|-1e31
        set Phase transition temperature upper limits = background:769.0|773.15|1e31|  1e31|1e31`
And each subgroup is separated with a blank from the others. In this way, it's apparent that the first 3 PTs are the basalt transitions at shallower depth, while the other 2 are PTs at deeper depth.

gassmoeller reviewed (approved):
Great, this looks good to go. I want to do some rewording for the changelog entry, but I will do that in a follow-up to not delay this PR further. Thanks for the work and sticking with us though all the discussions, I am glad we can merge this now.

lhy11009 commented:
Nice, thanks for your help and that brilliant modification of the phase
function implementation.

On Wed, Apr 26, 2023 at 1:58 PM Rene Gassmoeller ***@***.***>
wrote:

> Merged #4591 <https://github.com/geodynamics/aspect/pull/4591> into main.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/geodynamics/aspect/pull/4591#event-9107882996>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG2WMFJXJLVMCAMHL5I7IC3XDGEABANCNFSM5VWRJJNQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 

Ph.D student of Geophysics
Earth & Planetary Sciences Dept., UC Davis
Davis, CA 95616

mibillen commented:
Whohoo!!

:-)


> On Apr 26, 2023, at 2:05 PM, lhy11009 ***@***.***> wrote:
> 
> 
> Nice, thanks for your help and that brilliant modification of the phase
> function implementation.
> 
> On Wed, Apr 26, 2023 at 1:58 PM Rene Gassmoeller ***@***.***>
> wrote:
> 
> > Merged #4591 <https://github.com/geodynamics/aspect/pull/4591> into main.
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <https://github.com/geodynamics/aspect/pull/4591#event-9107882996>, or
> > unsubscribe
> > <https://github.com/notifications/unsubscribe-auth/AG2WMFJXJLVMCAMHL5I7IC3XDGEABANCNFSM5VWRJJNQ>
> > .
> > You are receiving this because you authored the thread.Message ID:
> > ***@***.***>
> >
> 
> 
> -- 
> 
> Ph.D student of Geophysics
> Earth & Planetary Sciences Dept., UC Davis
> Davis, CA 95616
> —
> Reply to this email directly, view it on GitHub <https://github.com/geodynamics/aspect/pull/4591#issuecomment-1524045034>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACS64UXLBSE5AWA2OVEOQ7DXDGEZFANCNFSM5VWRJJNQ>.
> You are receiving this because you were mentioned.
> 

____________________
Professor of Geophysics
Chair, EPS Graduate Program
Geophysics Minor Advisor

Department of Earth and Planetary Sciences
Univ. of California, Davis
Davis, CA 95616
2129 Earth & Physical Sciences Bldg.
Office Phone: (530) 752-4169
https://magalibillen.faculty.ucdavis.edu/

Students - need to meet with me? Check available appointment times: 
https://calendar.app.google/49asftepWVS1AFvy9 
Or e-mail me list of days/times you are available & topic of discussion.

Pronouns: She/her/hers
First Generation College Student: https://firstgen.ucdavis.edu/
